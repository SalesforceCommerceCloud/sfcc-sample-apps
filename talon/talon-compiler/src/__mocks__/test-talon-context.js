const path = require('path');
const { startContext, getContext, endContext } = require('../context/context-service');
const crypto = require("crypto");

jest.mock('../utils/observable-folder-hash');

const DEFAULT_OPTIONS = {
    templateDir: path.resolve(__dirname, '../__tests__/fixtures'),
    basePath: '/basePath',
    outputDir: '/'
};

async function startTestContext(options = {}) {
    return startContext({
        ...DEFAULT_OPTIONS,
        ...options
    }).then(context => {
        context.versionKey = options.versionKey || crypto.randomBytes(16).toString("hex");
    });
}

function updateTestContext(options) {
    Object.assign(getContext(), options);
}

module.exports = {
    startTestContext,
    getTestContext: getContext,
    updateTestContext,
    endTestContext: endContext,
    DEFAULT_OPTIONS
};
