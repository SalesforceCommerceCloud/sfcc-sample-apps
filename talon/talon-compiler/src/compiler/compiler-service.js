const { getContext, onContextClosed } = require('../context/context-service');
const { getOutputConfigs } = require('talon-common/dist/cjs');
const { resolveLwcNpmModules } = require('./module-resolver');
const LoadingCache = require('../utils/loading-cache');
const workerpool = require('workerpool');

// 1 pool of 2 minifier workers
const minifierPool = workerpool.pool(__dirname + '/minifier-worker.js', { minWorkers: 1, maxWorkers: 2 });

// 4 pools of 1 compiler workers, 1 per [env, target] pair
const compilerPools = new LoadingCache(() => {
    return workerpool.pool(__dirname + '/compiler-worker.js', { minWorkers: 1, maxWorkers: 1 });
});

onContextClosed(() => {
    minifierPool.terminate(true);
    compilerPools.getValues().forEach(pool => pool.terminate(true));
    compilerPools.invalidateAll();
});

// resolve LWC modules from NPM packages, this is done once and the result will be passed to the workers
const modulePaths = Object.entries(resolveLwcNpmModules()).reduce((paths, [id, { entry }]) => {
    paths[id] = entry;
    return paths;
}, {});

/**
 * Compile a module in all modes, bundling code as an AMD module.
 *
 * Note on virtual modules: LWC plugin uses the module path to
 * determine the module namespace and name, so be sure to include
 * these in the virtual module ids, see example below.
 *
 * @example
 * compile('./x/myModule/myModule.js', 'x/myModule', {
 *     './x/myModule/myModule.js': `import html from './myHtml.html'; export default { html };`,
 *     './x/myModule/myHtml.html': `<template><x-cmp1></x-cmp1></template>`,
 * })
 *
 * @param {string} input - the bundle's entry point, similar to Rollup's input option
 * @param {string} id - an ID to use for AMD/UMD bundles, similar to Rollup's output.amd.id
 * @param {object} virtualModules - A map of virtual modules to include when resolving modules during compilation
 * @returns {object} an object keyed by mode with the compiled code as values
 */
async function compile(input, id, virtualModules) {
    const context = getContext();
    const outputConfigs = getOutputConfigs(process.env.MODE && process.env.MODE.split(','));

    // get unique { env, target } for the outputConfigs
    const envsAndTargets = outputConfigs.filter(({ env, target }, index, self) => {
        return index === self.findIndex(other => other.env === env && other.target === target);
    });

    return Promise.all(envsAndTargets.map(({ env, target }) => {
        // compile the resources in the necessary env and target regardless of whether the resource should be minifed
        // minification happens later
        return compilerPools.get(`${env}:${target}`).proxy().then(worker => {
            return worker.compile({ input, id, virtualModules, target, env, context, modulePaths });
        }).then(content => {
            return { env, target, content };
        });
    })).then(contentEntries => {
        // minify the content when applicable
        return Promise.all(outputConfigs.map(async ({ mode, env, target, minify }) => {
            const { content } = contentEntries.find(entry => entry.env === env && entry.target === target);

            if (minify) {
                return minifierPool.proxy().then(worker => worker.minify({ input, target, env, content }).then(minifiedContent => {
                    return { mode, content: minifiedContent };
                }));
            }

            return { mode, content };
        }));
    }).then(contentEntries => {
        return contentEntries.reduce((contents, { mode, content }) => {
            contents[mode] = content;
            return contents;
        }, {});
    }).catch(err => {
        return Promise.reject(err);
    });
}

module.exports = { compile };
