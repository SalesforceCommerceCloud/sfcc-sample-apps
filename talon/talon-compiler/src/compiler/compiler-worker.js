const { compile } = require('./compiler');
const workerpool = require('workerpool');

workerpool.worker({
    compile
});
