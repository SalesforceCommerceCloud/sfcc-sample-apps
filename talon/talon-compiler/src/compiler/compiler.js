const { assert } = require('../utils/assert');
const { rollup } = require('rollup');
const { startContext, getContext, endContext } = require('../context/context-service');
const { verbose } = require('./utils');
const commonjs = require('rollup-plugin-commonjs');
const configService = require('../config/config-service');
const designtime = require('./rollup-plugin-designtime');
const labels = require('./rollup-plugin-labels');
const lwc = require('@lwc/rollup-plugin');
const memoizePlugin = require('./rollup-plugin-memoize');
const nodeResolve = require('rollup-plugin-node-resolve');
const progress = require('./rollup-plugin-progress');
const proxyCompat = require('rollup-plugin-compat');
const replace = require('rollup-plugin-replace');
const virtual = require('./rollup-plugin-virtual');

require('colors');

const EXTERNAL_NAMESPACES = [
    '@babel/runtime', // babel runtime helpers and regenerator are provided by es5-proxy-compat and included in talon.js (compat)
    '@salesforce/apex',
    '@salesforce/user',
    'proxy-compat', // provided by es5-proxy-compat and included in talon.js (compat)
    'talon'
];

const EXTERNAL_IDS = [
    '@salesforce/cssvars/customProperties',
    'assert',
    'aura-instrumentation',
    'aura-storage',
    'aura',
    'instrumentation/service',
    'lightning:IntlLibrary',
    'lightning/configProvider',
    'logger',
    'lwc',
    'wire-service'
];

const LWC_DEFAULT_OPTIONS = {
    stylesheetConfig: {
        customProperties: {
            resolution: { type: 'module', name: '@salesforce/cssvars/customProperties' }
        }
    },
    exclude: ['/**/*.mjs'],
    resolveFromSource: true,
    resolveFromPackages: false
};

/**
 * Returns true whether the module with the given id should be considered external.
 */
function external(externals, globals, id) {
    // module is considered external if it appears in talon.config.json `external`
    // but not in `globals` in which case we'll generate a virtual module for it
    if (externals.includes(id) && !globals.hasOwnProperty(id)) {
        return true;
    }

    return EXTERNAL_IDS.includes(id)
        || EXTERNAL_NAMESPACES.filter(ns => id.startsWith(`${ns}/`)).length > 0;
}

/**
 * Ignore CIRCULAR_DEPENDENCY warnings
 */
function onwarn(warning, warn) {
    if (warning.code !== 'CIRCULAR_DEPENDENCY') {
        warning.message = `[rollup] ${warning.message}`.yellow;
        warn(warning);
    }
}

function resolveModules(modulePaths) {
    return {
        name: 'rollup-plugin-resolve-modules',
        resolveId(id) {
            return modulePaths[id];
        }
    };
}

// modules cache used between rollup invocations
const modulesCache = {};

/**
 * Compile a module in the specified mode, bundling code as an AMD module.
 *
 * @param {object} config - the compile configuration
 * @param {string} config.input - the bundle's entry point, similar to Rollup's input option
 * @param {string} config.id - an ID to use for AMD/UMD bundles, similar to Rollup's output.amd.id
 * @param {object} config.virtualModules - A map of virtual modules to include when resolving modules during compilation
 * @param {object} config.target - the output target, es5 or es2017
 * @param {object} config.env - the output environment, development or production
 * @param {object} config.modulePaths - a map keyed by LWC module id pointing to file path, used to resolve LWC module ids for Rollup
 * @returns {string} the compiled code
 */
async function compile({ input, id, virtualModules = {}, target, env, context, modulePaths = {} }) {
    assert(input, 'input must be specified');
    assert(target, 'target must be specified');
    assert(env, 'env must be specified');

    await startContext(context);

    try {
        const { templateDir, isPreview } = getContext();
        // get Rollup external and globals config from talon.config.json
        const { rollup: { external: externals = [], output: { globals = {} } } } = configService.getConfig();

        // add one virtual module per global that will get the variable from the window object
        const globalsModules = Object.entries(globals).reduce((modules, [externalId, variableName]) => {
            modules[externalId] = `export default window[${JSON.stringify(variableName)}]`;
            return modules;
        }, {});

        // Memoize the LWC plugin.
        // This prevents lwcResolver.resolveLwcNpmModules() from being called multiple times
        // knowing that it takes around ~6 seconds locally when LGC modules are loaded.
        // It also memoizes plugin functions, mainly `transform()` which allows to reuse compiled modules.
        const memoizedLwc = memoizePlugin(lwc, { ...LWC_DEFAULT_OPTIONS, rootDir: templateDir });

        const rollupOptions = {
            input,
            external: external.bind(this, externals, globals),
            onwarn,
            plugins: [
                verbose && progress(`[${input}] (${target}, ${env})`),
                virtual({ ...virtualModules, ...globalsModules }),
                labels(),
                memoizedLwc,
                resolveModules(modulePaths),
                nodeResolve(),
                commonjs(),
                isPreview && designtime(),
                replace({ 'process.env.NODE_ENV': JSON.stringify(env) }),
                target === 'es5' && proxyCompat({ polyfills: false })
            ],
            cache: modulesCache[env] && modulesCache[env][target] && { modules: Object.values(modulesCache[env][target]) }
        };

        const bundle = await rollup(rollupOptions);

        // cache modules for reuse
        if (bundle.cache) {
            modulesCache[env] = modulesCache[env] || {};
            modulesCache[env][target] = modulesCache[env][target] || {};
            Object.assign(modulesCache[env][target], bundle.cache.modules.reduce((entries, module) => {
                entries[module.id] = module;
                return entries;
            }, {}));
        }

        const result = await bundle.generate({
            format: 'amd',
            amd: {
                id: id || input,
                define: 'Talon.moduleRegistry.define'
            }
        });

        return result.code;
    } finally {
        endContext();
    }
}

module.exports = { compile };