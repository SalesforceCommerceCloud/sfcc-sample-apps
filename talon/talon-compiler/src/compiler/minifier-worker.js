const { minify } = require('./minifier');
const workerpool = require('workerpool');

workerpool.worker({
    minify
});
