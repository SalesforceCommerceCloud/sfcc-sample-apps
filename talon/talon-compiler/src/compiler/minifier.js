const { assert } = require('../utils/assert');
const { clearAndWrite } = require('./utils');
const { performance } = require('perf_hooks');
const { verbose } = require('./utils');
const terser = require('terser');

/**
 * Minify code using terser.
 *
 * @param {object} config - the minification configuration
 * @param {string} config.input - the name of the resource being minified, this is for logging purpose
 * @param {string} config.target- the target of the resource being minified, this is for logging purpose
 * @param {string} config.env - the env of the resource being minified, this is for logging purpose
 * @param {string} config.content - the content to minify
 */
async function minify({ input, target, env, content }) {
    assert(input, 'input must be specified');
    assert(target, 'target must be specified');
    assert(env, 'env must be specified');
    assert(content, 'content must be specified');

    return new Promise((resolve, reject) => {
        const t0 = performance.now();
        const result = terser.minify(content);

        if (result.error) {
            reject(`Failed to minify ${input} (${target}, ${env}): ${result.error}`);
            return;
        }

        if (verbose) {
            clearAndWrite(`[${input}] (${target}, ${env}): Minified in ${Math.floor(performance.now() - t0)} ms\n`);
        }

        resolve(result.code);
    });
}

module.exports = { minify };