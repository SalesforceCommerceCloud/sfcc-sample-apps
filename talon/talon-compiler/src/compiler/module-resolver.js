/*
 * This code is mostly copied from https://github.com/salesforce/lwc/blob/master/packages/%40lwc/module-resolver/src/index.js
 * and uses grep to find LWC modules faster.
 *
 * We ignore .lwcrc as it does not seem to be used for now...
 */
const { performance } = require('perf_hooks');
const glob = require('glob');
const path = require('path');
const fs = require('fs');
const { execSync } = require('child_process');

require('colors');

const { log } = console;

const DEFAULT_NS = 'x';
const MODULE_EXTENSION = '.js';
const MODULE_ENTRY_PATTERN = `**/*${MODULE_EXTENSION}`;
const MODULE_PATH_IGNORE_PATTERN = '**/__*/**';

function loadLwcConfig(modulePath) {
    try {
        return require(path.join(modulePath, 'package.json')).lwc;
    } catch {
        return null;
    }
}

function resolveModulesInDir(fullPathDir) {
    return glob.sync(MODULE_ENTRY_PATTERN, { cwd: fullPathDir, ignore: MODULE_PATH_IGNORE_PATTERN }).reduce((mappings, file) => {
        const fileName = path.basename(file, MODULE_EXTENSION);
        const rootDir = path.dirname(file);
        // the glob library normalizes paths to forward slashes only - https://github.com/isaacs/node-glob#windows
        const rootParts = rootDir.split('/');
        const registry = {
            entry: path.join(fullPathDir, file),
            moduleSpecifier: null,
            moduleName: null,
            moduleNamespace: DEFAULT_NS
        };

        const dirModuleName = rootParts.pop();
        const dirModuleNamespace = rootParts.pop();
        if (dirModuleNamespace && dirModuleName === fileName) {
            registry.moduleName = fileName;
            registry.moduleNamespace = dirModuleNamespace.toLowerCase();
            registry.moduleSpecifier = `${dirModuleNamespace}/${fileName}`;
            mappings[registry.moduleSpecifier] = registry;
        }
        return mappings;
    }, {});
}

function hasModuleBeenVisited(module, visited) {
    if (visited.has(module)) {
        return true;
    }
    return false;
}

function resolveModules(modules, opts) {
    if (Array.isArray(modules)) {
        modules.forEach((modulePath) => resolveModules(modulePath, opts));
    } else {
        const { mappings, visited, moduleRoot, lwcConfig } = opts;
        if (typeof modules === 'string') {
            const packageEntries = resolveModulesInDir(path.join(moduleRoot, modules), lwcConfig);
            Object.keys(packageEntries).forEach((moduleName) => {
                if (!hasModuleBeenVisited(moduleName, visited)) {
                    mappings[moduleName] = packageEntries[moduleName];
                    visited.add(moduleName);
                }
            });
        } else {
            Object.keys(modules).forEach((moduleName) => {
                if (!hasModuleBeenVisited(moduleName, visited)) {
                    const modulePath = path.join(moduleRoot, modules[moduleName]);
                    mappings[moduleName] = { moduleSpecifier: moduleName, entry: modulePath };
                    visited.add(moduleName);
                }
            });
        }
    }
}

function findLwcModules(nodeModulesDir) {
    return execSync(`grep "\\"lwc\\"" ${nodeModulesDir} -R --include package.json -l  --exclude-dir @lwc --exclude-dir __tests__ || exit 0`)
            .toString()
            .split('\n')
            .map(path.dirname);
}

function resolveLwcNpmModules() {
    const t0 = performance.now();
    const visited = new Set();
    const mappings = {};

    for (const nodeModulesDir of module.paths.filter(fs.existsSync)) {
        const moduleRoots = findLwcModules(nodeModulesDir);
        for (const moduleRoot of moduleRoots) {
            const lwcConfig = loadLwcConfig(moduleRoot);

            if (lwcConfig) {
                resolveModules(lwcConfig.modules, {mappings, visited, moduleRoot, lwcConfig });
            }
        }
    }

    log(`Resolved ${Object.keys(mappings).length} LWC modules in ${Math.floor(performance.now() - t0)} ms`.green.dim);

    return mappings;
}

exports.resolveLwcNpmModules = resolveLwcNpmModules;
