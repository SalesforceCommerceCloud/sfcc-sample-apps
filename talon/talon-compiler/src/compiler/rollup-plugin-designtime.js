const path = require('path');

const frameworkModuleSrcDir = `${path.dirname(require.resolve('talon-framework/package.json'))}/src/modules`;
/**
 * Plugin that will resolve modules in talondesign namespace
 */
function designtime() {
    return {
        name: 'rollup-plugin-talon-designtime',

        resolveId(importee) {
            if (importee.startsWith('talondesign/')) {
                const [namespace, moduleName] = importee.split('/');
                return path.resolve(__dirname, `${frameworkModuleSrcDir}/${namespace}/${moduleName}/${moduleName}.js`);
            }
            return null;
        }
    };
}

module.exports = designtime;