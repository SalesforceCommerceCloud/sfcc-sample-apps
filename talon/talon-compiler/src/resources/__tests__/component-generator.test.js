import ComponentGenerator from '../component-generator';
import { startTestContext, endTestContext } from 'test-talon-context';
import { compile } from '../../compiler/compiler-service';

jest.mock('../../compiler/compiler-service');
jest.mock('../../metadata/metadata-service');

beforeEach(() => {
    jest.clearAllMocks();
    return startTestContext();
});

afterEach(() => {
    endTestContext();
});

const generator = new ComponentGenerator();

describe('component-generator', () => {
    compile.mockImplementation((componentName) => {
        return Promise.resolve({
            "dev": `dev code ${componentName}`,
            "prod": `prod code ${componentName}`
        });
    });

    it('generates component resource', async () => {
        return generator.get('component://x/cmp1').then((staticResource) => {
            expect(JSON.stringify(staticResource, null, 3)).toMatchSnapshot();
        });
    });

    it('compiles component', async () => {
        return generator.get('component://x/cmp1').then(() => {
            expect(compile.mock.calls).toHaveLength(1);
            expect(compile.mock.calls[0][0]).toBe('x/cmp1');
        });
    });
});

