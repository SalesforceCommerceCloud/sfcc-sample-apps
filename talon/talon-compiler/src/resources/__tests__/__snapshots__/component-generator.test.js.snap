// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component-generator generates component resource 1`] = `
"{
   \\"descriptor\\": \\"component://x/cmp1\\",
   \\"contents\\": {
      \\"dev\\": \\"dev code x/cmp1\\",
      \\"prod\\": \\"prod code x/cmp1\\"
   },
   \\"uids\\": {
      \\"dev\\": \\"4ca02f735e\\",
      \\"prod\\": \\"c8f94971a1\\"
   }
}"
`;
